Sources/MapboxCoreNavigation/Directions+NavigationProviderRequest.swift

```diff
@@ -0,0 +1,73 @@
+import MapboxDirections
+
+extension URLSessionDataTask: NavigationProviderRequest {
+    public var id: UInt64 {
+        UInt64(taskIdentifier)
+    }
+}
+
+extension Directions: NavigationProvider {
+    
+    @discardableResult public func requestRoutes(options: RouteOptions, completionHandler: @escaping RouteCompletionHandler) -> NavigationProviderRequest? {
+        return calculate(options, completionHandler: completionHandler)
+    }
+    
+    @discardableResult public func requestRoutes(options: MatchOptions, completionHandler: @escaping MatchCompletionHandler) -> NavigationProviderRequest? {
+        return calculate(options, completionHandler: completionHandler)
+    }
+    
+    @discardableResult public func refreshRoute(indexedRouteResponse: IndexedRouteResponse, fromLegAtIndex: UInt32, completionHandler: @escaping RouteCompletionHandler) -> NavigationProviderRequest? {
+        guard case let .route(routeOptions) = indexedRouteResponse.routeResponse.options else {
+            preconditionFailure("Invalid route data passed for refreshing.")
+        }
+        
+        let session = (options: routeOptions as DirectionsOptions,
+                       credentials: self.credentials)
+        
+        guard let responseIdentifier = indexedRouteResponse.routeResponse.identifier else {
+            defer {
+                completionHandler(session, .failure(.noData))
+            }
+            return nil
+        }
+        
+        return refreshRoute(responseIdentifier: responseIdentifier,
+                            routeIndex: indexedRouteResponse.routeIndex,
+                     fromLegAtIndex: Int(fromLegAtIndex),
+                     completionHandler: { credentials, result in
+                        switch result {
+                        case .failure(let error):
+                            completionHandler(session, .failure(error))
+                        case .success(let routeRefreshResponse):
+                            DispatchQueue.global().async {
+                                let routeResponse = indexedRouteResponse.routeResponse.copy(with: routeOptions)
+                                routeResponse.routes?[indexedRouteResponse.routeIndex].refreshLegAttributes(from: routeRefreshResponse.route)
+                                DispatchQueue.main.async {
+                                    completionHandler(session, .success(routeResponse))
+                                }
+                            }
+                        }
+                     })
+    }
+}
+
+extension RouteResponse {
+    func copy(with options: DirectionsOptions) -> RouteResponse {
+        var copy = self
+        copy.routes = copy.routes?.map { $0.copy(with: options) }
+        return copy
+    }
+}
+
+extension Route {
+    func copy(with options: DirectionsOptions) -> Route {
+        let encoder = JSONEncoder()
+        encoder.userInfo[.options] = options
+        let encoded = try! encoder.encode(self)
```

Comment:
Avoid these implicitly unwrapped optionals, especially because the DirectionsOptions object could be a subclass of RouteOptions defined by the developer, who may need to catch an encoding error. For consistency with #3484, this method should be throwable.